File: vectorize.py
import os
import re
import time
import requests
from pathlib import Path
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent / ".env")

VECTORIZE_URL = os.getenv("VECTORIZE_URL", "").rstrip("/")
API_KEY = os.getenv("VECTORIZE_API_KEY", "")

HEADERS = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json",
}


def _clean(text: str) -> str:
    text = re.sub(r'\s*https://t\.co/\S+', '', text)
    return text.replace("\n", " ").strip()


def ingest_batch(tweets: list[dict]) -> tuple[int, int]:
    """Send up to 100 tweets in one batch. Returns (succeeded, failed)."""
    documents = []
    for tweet in tweets:
        content = _clean(tweet.get("text", ""))
        if not content or len(content) < 10:
            continue
        documents.append({
            "id": f"tweet_{tweet['id']}",
            "content": content,
            "metadata": {
                "author":     tweet.get("author_username", ""),
                "name":       tweet.get("author_name", ""),
                "likes":      int(tweet.get("likes") or 0),
                "retweets":   int(tweet.get("retweets") or 0),
                "views":      int(tweet.get("views") or 0),
                "url":        tweet.get("url", ""),
                "created_at": (tweet.get("created_at") or "")[:10],
                "source":     "bookmark",
            },
        })

    if not documents:
        return 0, len(tweets)

    try:
        r = requests.post(
            f"{VECTORIZE_URL}/ingest/batch",
            json={"documents": documents, "concurrency": 10},
            headers=HEADERS,
            timeout=120,
        )
        r.raise_for_status()
        data = r.json()
        return data.get("succeeded", 0), data.get("failed", 0)
    except Exception as e:
        return 0, len(documents)


def semantic_search(query: str, limit: int = 20) -> list[dict]:
    """Semantic search via the vectorize worker."""
    payload = {
        "query": query,
        "topK": min(limit, 50),
        "includeMetadata": True,
        "rerank": True,
    }
    try:
        r = requests.post(
            f"{VECTORIZE_URL}/search",
            json=payload,
            headers=HEADERS,
            timeout=20,
        )
        r.raise_for_status()
        data = r.json()
        return data.get("results", data.get("matches", []))
    except Exception as e:
        raise RuntimeError(f"Search failed: {e}")

File: telegram-bot/src/vectorize.js
const MAX_SCORE = 2.0;
const VECTORIZE_INTERNAL_URL = "https://vectorize-mcp-worker.fpl-test.workers.dev/search";

export async function searchCorpus(query, { vectorizeWorker, vectorizeApiKey, limit = 15 }) {
  const res = await vectorizeWorker.fetch(VECTORIZE_INTERNAL_URL, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${vectorizeApiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      query,
      topK: Math.min(limit, 50),
      includeMetadata: true,
      rerank: true,
    }),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Vectorize search failed: ${res.status} — ${body.slice(0, 200)}`);
  }

  const data = await res.json();
  const results = data.results ?? data.matches ?? [];
  return results.filter(r => (r.score ?? 0) <= MAX_SCORE);
}

export function filterByAuthors(results, targets) {
  if (!targets || targets.length === 0) return [];
  const normalized = targets.map(t => t.toLowerCase().replace(/^@/, ""));
  return results.filter(r => {
    const author = (r.metadata?.author ?? "").toLowerCase().replace(/^@/, "");
    return normalized.includes(author);
  });
}

export function fitScore(results, targets) {
  if (!results.length) return 0;
  const matches = filterByAuthors(results, targets);
  return Math.round((matches.length / results.length) * 100);
}

export function formatResult(r) {
  const text = (r.text ?? r.content ?? "").replace(/\s*https?:\/\/\S+/g, "").trim();
  const tweetId = (r.id ?? "").replace(/^tweet_/, "").replace(/-chunk-\d+$/, "");
  const url = r.metadata?.url || (tweetId ? `https://x.com/i/web/status/${tweetId}` : "");
  return {
    author: r.metadata?.author ?? null,
    text: text.length > 200 ? text.slice(0, 200) + "…" : text,
    likes: r.metadata?.likes ?? null,
    url,
    score: r.score ?? 0,
  };
}
